File: ShibuyaTools/formats/Xml.fs
module Xml

type Translation =
    { Key: string
      Val: string }

type Bookmark =
    { Location: int32
      LocationOffset: int64 }

let private readBookmarkList scriptOffset reader =
    let bookmarkList = ResizeArray()

    while Reader.tell reader < scriptOffset do
        let _ = Reader.readInt32 reader // kind
        let locationOffset = Reader.tell reader
        let location = Reader.readInt32 reader
        let _ = Reader.readNullTerminatedString reader // name
        Reader.align reader
        bookmarkList.Add({ Location = location; LocationOffset = locationOffset })

    bookmarkList.Sort(fun x y -> x.Location - y.Location)
    bookmarkList

let extract path body =
    use reader = Reader.fromByteArray body
    Reader.seekFromStart 1072 reader
    let scriptOffset = Reader.readInt32 reader
    let bookmarkList = readBookmarkList scriptOffset reader
    let mutable sourcePtr = scriptOffset
    let translations = ResizeArray()

    let registerBlock body =
        let mutable ptr = 0
        for (_, b, key) in Phrases.detect body do
            translations.Add({ Key = key; Val = key })
            ptr <- b
        sourcePtr <- sourcePtr + body.Length

    for bookmark in bookmarkList do
        let offset = scriptOffset + bookmark.Location
        registerBlock body[sourcePtr..offset-1]

    let updated = List.ofSeq translations
    let source = Json.read [] path
    if updated = source then () else
    Json.write updated path

let translate path body =
    let mutable index = 0
    let mutable existing = Json.read [||] path
    let mutable showError = true

    let popval key =
        if index >= existing.Length then key else
        let i = index
        index <- i + 1
        let e = existing[i]
        if e.Key = key then e.Val else
        if showError then eprintfn "%s: wrong key [%d] '%s'; expected '%s'" path i e.Key key
        showError <- false
        key

    use reader = Reader.fromByteArray body
    Reader.seekFromStart 1072 reader
    let scriptOffset = Reader.readInt32 reader
    let bookmarkList = readBookmarkList scriptOffset reader
    let mutable sourcePtr = scriptOffset
    let mutable targetPtr = scriptOffset
    let blocks = ResizeArray()
    let bookmarkBlock = body[..scriptOffset-1]
    use bookmarkWriter = Writer.fromByteArray bookmarkBlock
    blocks.Add(bookmarkBlock)

    let registerBlock body =
        let chunks = ResizeArray()
        let mutable ptr = 0
        for (a, b, phrase) in Phrases.detect body do
            let translation = popval phrase
            chunks.Add(body[ptr..a-1])
            chunks.Add(Phrases.encode translation)
            ptr <- b
        chunks.Add(body[ptr..])
        let block = Array.concat chunks
        blocks.Add(block)
        sourcePtr <- sourcePtr + body.Length
        targetPtr <- targetPtr + block.Length

    for bookmark in bookmarkList do
        let offset = scriptOffset + bookmark.Location
        registerBlock body[sourcePtr..offset-1]
        Writer.seekFromStart bookmark.LocationOffset bookmarkWriter
        Writer.writeInt32 (targetPtr - scriptOffset) bookmarkWriter

    let index = Array.findIndexBack ((=) 0xACuy) body
    registerBlock body[sourcePtr..index+1]
    let totalLength = blocks |> Seq.sumBy (fun x -> x.Length)
    let alignOffset = totalLength % 4
    if alignOffset <> 0 then blocks.Add(Array.zeroCreate (4 - alignOffset))
    Array.concat blocks

File: ShibuyaTools.Resources.Wad/Xml.cs
using Microsoft.Extensions.Logging;
using ShibuyaTools.Core;

namespace ShibuyaTools.Resources.Wad;

internal abstract class Xml
{
    public static readonly Xml Sns = new XmlSns();
    public static readonly Xml Sns64 = new XmlSns64();

    private record Bookmark(int Location, long LocationOffset);

    public XmlTranslation[] Parse(byte[] body)
    {
        using var stream = new MemoryStream(body);
        using var reader = new BinaryReader(stream);

        var scriptOffset = (int)ReadScriptOffset(reader);
        var bookmarkList = ReadBookmarkList(reader, scriptOffset);

        var sourcePtr = scriptOffset;
        var translations = new List<XmlTranslation>();

        foreach (var bookmark in bookmarkList)
        {
            var offset = scriptOffset + bookmark.Location;
            RegisterBlock(body[sourcePtr..offset]);
        }

        RegisterBlock(body[sourcePtr..]);

        return [.. translations];

        void RegisterBlock(byte[] body)
        {
            var ptr = 0;

            foreach (var (_, b, key) in Phrases.Detect(body))
            {
                translations.Add(new XmlTranslation(Key: key, Val: key));
                ptr = b;
            }

            sourcePtr += body.Length;
        }
    }

    public ReadOnlySpan<byte> Translate(ILogger logger, XmlTranslation[] existing, ReadOnlySpan<byte> body)
    {
        var index = 0;
        var showError = true;

        string popval(string key)
        {
            if (index >= existing.Length)
            {
                logger.LogError("translation eos at key '{key}'", key);
                return key;
            }

            var i = index++;
            var e = existing[i];
            if (e.Key == key)
            {
                return XmlText.Compact(e.Val);
            }

            if (showError)
            {
                logger.LogError("wrong key [{index}] '{actualKey}'; expected '{expectedKey}'", i, e.Key, key);
                showError = false;
            }

            return key;
        }

        using var stream = new MemoryStream(body.Length);
        using var reader = new BinaryReader(stream);

        stream.Write(body);

        var scriptOffset = (int)ReadScriptOffset(reader);
        var bookmarkList = ReadBookmarkList(reader, scriptOffset);

        var sourcePtr = scriptOffset;
        var targetPtr = scriptOffset;

        var blocks = new List<byte[]>();
        var bookmarkBlock = body[..scriptOffset].ToArray();

        using var bookmarkStream = new MemoryStream(bookmarkBlock);
        using var bookmarkWriter = new BinaryWriter(bookmarkStream);

        blocks.Add(bookmarkBlock);

        void RegisterBlock(ReadOnlySpan<byte> body)
        {
            var block = new List<byte>(body.Length);
            var ptr = 0;

            foreach (var (a, b, phrase) in Phrases.Detect(body.ToArray()))
            {
                var translation = popval(phrase);
                block.AddRange(body[ptr..a]);
                block.AddRange(Phrases.Encode(translation));
                ptr = b;
            }

            block.AddRange(body[ptr..]);
            blocks.Add([.. block]);
            sourcePtr += body.Length;
            targetPtr += block.Count;
        }

        foreach (var bookmark in bookmarkList)
        {
            var offset = scriptOffset + bookmark.Location;
            RegisterBlock(body[sourcePtr..offset]);
            bookmarkStream.Position = bookmark.LocationOffset;
            bookmarkWriter.Write(targetPtr - scriptOffset);
            bookmarkWriter.Flush();
        }

        RegisterBlock(body[sourcePtr..].TrimEnd((byte)0));

        var totalLength = blocks.Sum(b => b.Length);
        var alignment = totalLength % 4;

        if (alignment != 0)
        {
            totalLength += 4 - alignment;
        }

        var result = new byte[totalLength];
        var p = 0;

        foreach (var block in blocks)
        {
            block.CopyTo(result, p);
            p += block.Length;
        }

        return result;
    }

    protected abstract long ReadScriptOffset(BinaryReader reader);

    private static List<Bookmark> ReadBookmarkList(BinaryReader reader, long scriptOffset)
    {
        var stream = reader.BaseStream;
        var bookmarkList = new List<Bookmark>();

        while (stream.Position < scriptOffset)
        {
            _ = reader.ReadInt32(); // kind
            var locationOffset = stream.Position;
            var location = reader.ReadInt32();
            _ = reader.ReadNullTerminatedString(); // name

            stream.Align();

            bookmarkList.Add(
                new Bookmark(
                    Location: location,
                    LocationOffset: locationOffset));
        }

        bookmarkList.Sort((a, b) => a.Location - b.Location);

        return bookmarkList;
    }

    private class XmlSns : Xml
    {
        protected override long ReadScriptOffset(BinaryReader reader)
        {
            reader.BaseStream.Position = 0x430;

            return reader.ReadInt32();
        }
    }

    private class XmlSns64 : Xml
    {
        protected override long ReadScriptOffset(BinaryReader reader)
        {
            reader.BaseStream.Position = 0x830;

            return reader.ReadInt64();
        }
    }
}
